### ISSUE TITLE
Improve the UX of Visual Scripting Nodes

### ISSUE THREAD

**Issue by swarnimarun** [godotengine/godot-proposals#5]
List Type Nodes
---
- Allow the Node Graph to be more self-sufficient and not have to move to the inspector for every new change.
- Make nodes that use color codes with more rigidity and have constant theme.
- More Dynamic Nodes, nodes that can change and are well contained for the purpose.

![image](https://user-images.githubusercontent.com/19930870/57286762-9b742000-70d3-11e9-8305-434d564a0483.png)

> This is a quick MockUp to help understand the idea.

Here the color code is that brighter color is for the Specific Type and then the duller(whitish color) is for the Array of that type.
- In this Grey is Any, Whitish Grey is Array(of all types).
- Yellow is Int, Light Yellow is for Int Array(PoolIntArray).
- Blue is for Object.

__We have + and x buttons to add and remove elements.__

- For arrays it will just allow us to select a type and create a new variable.
- While for setter fields they will pop open a menu to select a field we want to edit(as not all fields need to be changed)

__This will make the setter node much more powerful and much less redundant(each setter editing one value was kinda wonky) and we obviously need a create Array node.__

This design theme can be followed to other aspects of the Visual Script system as well. Like Getters, Maps and etc. 

Probably have even the **Curve** and **Gradient Editors** be a part of the Visual Script directly to make editing fields of even resources easier.

For now we can add this as extra type of nodes to allow for backwards compatibility.

**Comment by swarnimarun** [comment]
cc: @fire

**Comment by fire** [comment]
@swarnimarun Can you update this with you pr number?

**Comment by VicentGJ** [comment]
Looks nice, it's a really needed change before doing practically anything else in VS system.
For now it looks so bad to the eye, and the differentiation between different kinds of nodes is not clear enough

**Comment by lentsius-bark** [comment]
Yes please! Making the nodes self sufficient without having to leave them for every change is an awesome UX change.

**Comment by swarnimarun** [comment]
Here's the commit that does most of the foundational work for the changes,
https://github.com/godotengine/godot/pull/29681

Added some of the basic nodes for some of the features mentioned above and some of the UX changes.

I will make another draft of ideas for some of the visual changes, feel free to provide some advice that you might believe would be a better look for VisualScript. Although I am working mostly on features rather than the Theme of the Visual Script itself.

**Comment by Norrox** [comment]
is this possible to add to the visual shader to? if so then maybe my proposal is not needed ? :P  https://github.com/godotengine/godot-proposals/issues/134

**Comment by swarnimarun** [comment]
@norrox yes it's possible to add similar feature to Visual Shader but it would need to be rewritten and the UI changes still haven't been made.

**Comment by Gianclgar** [comment]
There are more ideas that could be useful for this at #506 and #593

**Comment by Norrox** [comment]
> Here's the commit that does most of the foundational work for the changes,
> [godotengine/godot#29681](https://github.com/godotengine/godot/pull/29681)
> 
> Added some of the basic nodes for some of the features mentioned above and some of the UX changes.
> 
> I will make another draft of ideas for some of the visual changes, feel free to provide some advice that you might believe would be a better look for VisualScript. Although I am working mostly on features rather than the Theme of the Visual Script itself.

I have always felt that Bolt has a nice Visual Script theme. Just to put some ideas here 😊
![Bolt5](https://user-images.githubusercontent.com/13393087/88870442-a9045100-d215-11ea-9737-392bce5403a0.gif)
![Messy](https://user-images.githubusercontent.com/13393087/88870452-adc90500-d215-11ea-90f6-3a42f05caa8a.jpg)
This os from bolt 2
![e568645d-3f63-47f1-9feb-f564b787dcde_Bolt2Mockup](https://user-images.githubusercontent.com/13393087/88870473-bb7e8a80-d215-11ea-8e66-672567bf21ca.png)
